Feature fee center

lomamech opened the pull request:
This PR reconstructs the modeling of the Fee center analysis all, this first step allows to define a Fee Center while specifying if it is a Principal or auxiliary center,
 - for the Principal center it is necessary to assign a reference of Profit account and a reference for the cost 
- for the auxiliary center it is necessary to specify whether it is a cost center or a profit center, 

There is also the possibility of assigning a cost center to a Project or several, for the account reference we had to reuse the one that already exists in the system

- Implement API for Fees Centers
- Implement Client Side for Fees Centers
- Add components for Accounts References and Select Multiples Projects
- Implementation of integration test for Fee Center Manager
- Update and Improve client side for Fee Center
- Implement Fee Center E2E test
- Set Data for Database for test
- Update E2E test for Account-Reference
- Add in shared Components for bhAccountReferenceSelect and

closes #3006

lomamech committed:
feature(Fees Centers Modules)
- Implement API for Fees Centers
- Implement Client Side for Fees Centers
- Add components for Accounts References and Select Multiples Projects

lomamech committed:
test(Fee Center)
- Implementation of integration test for Fee Center Manager
- Update and Improve client side for Fee Center

lomamech committed:
test(Fee Center)
- Implement Fee Center E2E test
- Set Data for Database for test
- Update E2E test for Account-Reference
- Add in shared Components for bhAccountReferenceSelect and
  bhProjectMultipleSelect

lomamech committed:
improvement(Fee Center)
- Improve and sanitaze code

lomamech committed:
improvement(Fee Center) Sanitaze Code

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -510,6 +513,7 @@
             "PRICE": "Price",
             "PRICE_LIST": "Price List",
             "PRINCIPAL": "Principal",
+            "PRINCIPAL_2": "Principal",
```
What is the difference between "PRINCIPAL" and "PRINCIPAL_2"?  Can we delete "PRINCIPAL_2"?

lomamech replied:
The difference is in French, the word fee Center is of the masculine gender, hence the qualifier "Principal" should not have an "e" at the end, which is why I added PRINCIPAL_2 which in the translation French is devoid of "e"

jniles commented on client/src/js/components/bhAccountReferenceSelect.js:
```diff
@@ -0,0 +1,64 @@
+      required         : '<?',
+      label            : '@?',
+      name             : '@?',
+      validationTrigger :  '<?',
```
With the update to Angular 1.7, we no longer need to pass `validationTrigger` into the component.  For more discussion, you can see some of it in https://github.com/IMA-WorldHealth/bhima-2.X/pull/3064.

Instead, you can just replace all calls to `$ctrl.validationTrigger` with `AccountReferenceForm.$submitted`.  That way, it will run validation when the form is submitted.

jniles replied:
Just made an issue about this (#3134).

jniles commented on client/src/modules/fee_center/fee_center.html:
```diff
@@ -0,0 +1,45 @@
+        </div> 
+        
+        <div class="toolbar-item"> 
+          <button 
```
This should use our `bhFilterToggle` component.

jniles commented on client/src/modules/fee_center/fee_center.html:
```diff
@@ -0,0 +1,45 @@
+      
+      <div class="toolbar"> 
+        <div class="toolbar-item"> 
+          <button class="btn btn-default text-capitalize" id="create" ui-sref="fee_center.create" data-method="create"> 
```
It doesn't look like you ever use `id="create"`.  Could you remove it?

jniles commented on client/src/modules/fee_center/fee_center.js:
```diff
@@ -0,0 +1,101 @@
+        displayName : '',
+        headerCellFilter : 'translate',
+        enableFiltering : false,
+        enableSorting : false,
```
It might be nice to sort on `is_principal`.  It would be an easy way to group them together at the top or bottom of the grid.

jniles commented on client/src/modules/fee_center/fee_center.js:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  // create a new FeeCenter
+  function createFeeCenter() {
```
This function isn't used - can it be removed?

lomamech committed:
Update Feature Fee Center

jniles commented on client/src/modules/fee_center/fee_center.service.js:
```diff
@@ -0,0 +1,17 @@
+ * Encapsulates common requests to the /fee_center/ URL.
+ */
+function FeeCenterService(Api) {
+  const service = new Api('/fee_center/');
```
👍

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.html:
```diff
@@ -0,0 +1,176 @@
+      </div>
+    </div>
+
+    <div class="form-group">
```
Is one of these choices required?  If so, this should have validation on it.

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.html:
```diff
@@ -0,0 +1,176 @@
+          on-select-callback="FeeCenterModalCtrl.onSelectAccountReference(accountReference, 0)"
+          required="true"
+          validation-trigger="FeeCenterForm.$submitted">
+        </bh-account-config-select>
```
This component seems broken - you start with `bh-account-reference-select`, but you end with `</bh-account-config-select>`.

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.html:
```diff
@@ -0,0 +1,176 @@
+        <hr />
+      </span>
+
+      <div ng-if="FeeCenterModalCtrl.feeCenter.is_principal" class="checkbox">
```
It looks like you do this check: `FeeCenterModalCtrl.feeCenter.is_principal` multiple times.  It would be nicer to group this under a larger section, something like:
```html
<section ng-if="FeeCenterModalCtrl.feeCenter.is_principal">
  <hr />

  <div class="checkbox">
    <! -- ... etc ... -->
  </div>
</section>
```

lomamech replied:
It's just because I use the same <bh-account-reference-select> as it is for the case of the Principal Fee Center or the auxiliary centers (Costs or Profits)

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.html:
```diff
@@ -0,0 +1,176 @@
+          on-select-callback="FeeCenterModalCtrl.onSelectAccountReference(accountReference, 1)"
+          required="true"
+          validation-trigger="FeeCenterForm.$submitted">
+        </bh-account-config-select>
```
Same issue with the component here.

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.html:
```diff
@@ -0,0 +1,176 @@
+      </div>
+
+      <div ng-if="(FeeCenterModalCtrl.feeCenter.is_principal && FeeCenterModalCtrl.hasCostCenter) || (FeeCenterModalCtrl.auxiliaryCenter && FeeCenterModalCtrl.isCostCenter)">
+        <bh-account-reference-select
```
Why is a cost center an account reference?  I thought cost centers were different from accounts...

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.html:
```diff
@@ -0,0 +1,176 @@
+        </span>
+      </label>
+    </div>
+    <div ng-if="FeeCenterModalCtrl.assignedProjects">
```
Why does a cost center need to know about a project?

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.js:
```diff
@@ -0,0 +1,154 @@
+
+  // exposed methods
+  vm.submit = submit;
+  vm.closeModal = closeModal;
```
I don't think `closeModal()` is ever called or used.

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.js:
```diff
@@ -0,0 +1,154 @@
+          vm.projects = data.projects;
+        }
+
+        if (data.references) {
```
This has a lot of nesting!  This is known as [callback hell](http://callbackhell.com/) and isn't great practice.

Let's make this easier to deal with by moving most of this code into functions.  We'll make the following functions:
```js
function processProjects(data) { }

function processReference(reference) { }
```

Inside `processReference()` you should put all the code from the `forEach()` loop.  The `processProjects()` will contain the code from the `if (data.projects)` statement.

jniles commented on server/controllers/finance/feeCenter.js:
```diff
@@ -0,0 +1,244 @@
+  const dataReferences = [];
+  const dataProjects = [];
+
+  if (data.reference_fee_center) {
```
This code would be easier to read if you grouped all the the common operations together.  Do all things related to `references` first, then do all things related to project, etc...  Like this:
```js
if (data.reference_fee_center) {
  const dataReferences =  data.reference_fee_center.map(item => [
    feeCenterId,
    item.account_reference_id,
    item.is_cost,
  ]);

  const sqlReferences = `
      INSERT INTO reference_fee_center (fee_center_id, account_reference_id, is_cost) VALUES ?`;

 transaction
      .addQuery(sqlReferences, [dataReferences]);
}
```

That way, the developer doesn't have to think about all of the variables all the time,  You could even move this code into a function like:
```js
function processReferences(data, transaction) { /* the code above */ }

if (data.reference_fee_center.length) {
  processReferences(data, transaction);
}
```

jniles reviewed (commented):
A few cleanup comments on this code.  Some parts of it are pretty hard to read.

You should show this to Mr. Wayne before we get too far into the logic of how fee centers should work.  I'd be interested in having his feedback on this.

lomamech commented on client/src/i18n/fr/form.json:
```diff
@@ -508,6 +511,7 @@
             "PRICE": "Prix",
             "PRICE_LIST": "Liste de Prix",
             "PRINCIPAL": "Principale",
+            "PRINCIPAL_2": "Principal",
```
@jniles  this the difference

lomamech committed:
refactor(Fee Center)
- Sanitaze code and remove unused property, functions and id
- Use bh-filter-toogle
- Enable Sorting in fee center grid
- Invalidate form Error and display messages
- Deletion of assignment of cost centers to projects

lomamech committed:
improvement(Fee Center)
- Improvement of Fee Center Management, Sanitaze and Cleaning Code

jniles commented on client/src/js/components/bhProjectsMultipleSelect.js:
```diff
@@ -0,0 +1,47 @@
+    $ctrl.label = $ctrl.label || 'FORM.LABELS.PROJECT';
+
+    // fired when the Projects has been selected or removed from the list
+    $ctrl.onChange = $ctrl.onChange;
```
There isn't any need to do this - in fact, this line doesn't do anything.  It is like putting `1 = 1`.  Go ahead and remove it.

jniles reviewed (commented):
This branch is still actively being worked on, so I'll merge it into a feature branch on upstream.  I'll be interested to see how it works!
